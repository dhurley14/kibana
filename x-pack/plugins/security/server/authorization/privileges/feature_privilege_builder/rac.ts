/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { uniq } from 'lodash';

import type { FeatureKibanaPrivileges } from '../../../../../features/server';
import { BaseFeaturePrivilegeBuilder } from './feature_privilege_builder';

const readOperations: string[] = ['get', 'find'];
const writeOperations: string[] = ['update'];
const allOperations: string[] = [...readOperations, ...writeOperations];

export class FeaturePrivilegeRacBuilder extends BaseFeaturePrivilegeBuilder {
  public getActions(privilegeDefinition: FeatureKibanaPrivileges, spaceId: string): string[] {
    const getRacPrivilege = (owners: readonly string[], operations: string[]) => {
      return owners.flatMap((owner) =>
        operations.map((operation) => this.actions.rac.get(spaceId, owner, operation))
      );
    };

    return uniq([
      ...getRacPrivilege(allOperations, privilegeDefinition.alerts?.all ?? []),
      ...getRacPrivilege(readOperations, privilegeDefinition.alerts?.read ?? []),
    ]);
  }
}
