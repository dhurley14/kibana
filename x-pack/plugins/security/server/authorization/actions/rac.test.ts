/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { RacActions } from './rac';

const version = '1.0.0-zeta1';

describe('#get', () => {
  [null, undefined, '', 1, true, {}].forEach((spaceId: any) => {
    test(`spaceId of ${JSON.stringify(spaceId)} throws error`, () => {
      const racActions = new RacActions(version);
      expect(() =>
        racActions.get(spaceId, 'consumer', 'foo-action')
      ).toThrowErrorMatchingSnapshot();
    });
  });

  [null, undefined, '', 1, true, {}].forEach((operation: any) => {
    test(`operation of ${JSON.stringify(operation)} throws error`, () => {
      const racActions = new RacActions(version);
      expect(() =>
        racActions.get('foo-alertType', 'consumer', operation)
      ).toThrowErrorMatchingSnapshot();
    });
  });

  [null, '', 1, true, undefined, {}].forEach((consumer: any) => {
    test(`consumer of ${JSON.stringify(consumer)} throws error`, () => {
      const racActions = new RacActions(version);
      expect(() =>
        racActions.get('foo-alertType', consumer, 'operation')
      ).toThrowErrorMatchingSnapshot();
    });
  });

  test('returns `rac:${spaceId}:${owner}/${operation}`', () => {
    const racActions = new RacActions(version);
    expect(racActions.get('foo-spaceId', 'owner', 'bar-operation')).toBe(
      'rac:1.0.0-zeta1:foo-spaceId:owner/bar-operation'
    );
  });
});
