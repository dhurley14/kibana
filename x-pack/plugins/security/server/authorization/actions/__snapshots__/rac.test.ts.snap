// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#get consumer of "" throws error 1`] = `"\\"owner\\" is required and must be a string"`;

exports[`#get consumer of {} throws error 1`] = `"\\"owner\\" is required and must be a string"`;

exports[`#get consumer of 1 throws error 1`] = `"\\"owner\\" is required and must be a string"`;

exports[`#get consumer of null throws error 1`] = `"\\"owner\\" is required and must be a string"`;

exports[`#get consumer of true throws error 1`] = `"\\"owner\\" is required and must be a string"`;

exports[`#get consumer of undefined throws error 1`] = `"\\"owner\\" is required and must be a string"`;

exports[`#get operation of "" throws error 1`] = `"\\"operation\\" is required and must be a string"`;

exports[`#get operation of {} throws error 1`] = `"\\"operation\\" is required and must be a string"`;

exports[`#get operation of 1 throws error 1`] = `"\\"operation\\" is required and must be a string"`;

exports[`#get operation of null throws error 1`] = `"\\"operation\\" is required and must be a string"`;

exports[`#get operation of true throws error 1`] = `"\\"operation\\" is required and must be a string"`;

exports[`#get operation of undefined throws error 1`] = `"\\"operation\\" is required and must be a string"`;
